### DISCUSSION TITLE
Native Class-like Enumerations

### DISCUSSION THREAD

**Discussion by john-larson** [dotnet/csharplang#5122] [Language Ideas]
Enums in C# do not have object oriented features such as inheritance or methods. And this leads to code smells as described in:
https://docs.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types.

To overcome these shortcomings, custom enumeration classes are used as described in the aforementioned guide by Microsoft. 

This pattern is also implemented by a popular library: https://github.com/ardalis/SmartEnum.

But the problem with using this pattern is, because it is not a native type, serialization/deserialization requires further customizations in the components serializing these classes as enums.

SmartEnum library I mentioned contains some packages to overcome this issue in some of the external libraries but it is far from being exhaustive. Just as a basic example, a converter for OpenApi does not exist.

If C# gets to have a native enumeration class, this pattern will be standardized so that any serialization/deserialization provider will by default include it in their package. 

It will also enable a much cleaner syntax for the pattern.

And it looks like a sought after feature by many: https://stackoverflow.com/questions/757684/enum-inheritance

**Comment by DavidArno** [comment]
[Oracle own a patent on OO enums, that doesn't expire until 2024](https://patents.google.com/patent/US7263687), so Microsoft's hands are tied here.

Further, [the proposal for discriminated unions](https://github.com/dotnet/csharplang/blob/main/proposals/discriminated-unions.md) covers many use cases for expanded enum functionality without infringing on that patent and is therefore more likely to be the C# solution to this space.

**Reply by redradist** [reply]
@DavidArno 
> 
> 
> [Oracle own a patent on OO enums, that doesn't expire until 2024](https://patents.google.com/patent/US7263687), so Microsoft's hands are tied here.

Really !!!!!!!!
I did not believe until I saw it ... What a strange world ...
